x-pack/plugins/triggers_actions_ui/public/application/components/add_message_variables.tsx

```diff
@@ -107,15 +105,130 @@ export const AddMessageVariables: React.FunctionComponent<Props> = ({
     return <></>;
   }
 
+  const ToolTipContent = ({ description, label }: { description: string; label: string }) => {
+    return (
+      <>
+        <EuiText
+          size="s"
+          style={{
+            fontWeight: euiTheme.font.weight.bold,
+          }}
+        >
+          {label}
+        </EuiText>
+        <EuiSpacer size="s" />
+        <hr />
+        <EuiSpacer size="s" />
+        <EuiText size="xs">{description}</EuiText>
+      </>
+    );
+  };
+
+  const renderOption = (
+    option: EuiSelectableOption<{ description?: string }>,
+    searchValue: string
+  ) => {
+    return (
+      <EuiFlexGroup data-test-subj={`variableMenuButton-${option.label}`}>
+        <EuiFlexItem>
+          <EuiText
+            size="s"
+            style={{
+              fontWeight: euiTheme.font.weight.bold,
+            }}
+          >
+            <EuiHighlight search={searchValue}>{option.label}</EuiHighlight>
+          </EuiText>
+          <EuiSpacer size="xs" />
+          {option.description && (
+            <EuiText size="xs" color="subdued">
+              <EuiToolTip
+                display="block"
+                position="top"
+                content={<ToolTipContent description={option.description} label={option.label} />}
+                data-test-subj={`${option.label}-tooltip`}
+              >
+                <TruncatedText text={option.description || ''} />
+              </EuiToolTip>
+            </EuiText>
+          )}
+        </EuiFlexItem>
+      </EuiFlexGroup>
+    );
+  };
+
   return (
     <EuiPopover
       button={Button}
       isOpen={isVariablesPopoverOpen}
       closePopover={() => setIsVariablesPopoverOpen(false)}
-      panelPaddingSize="none"
-      anchorPosition="downLeft"
+      panelPaddingSize="s"
+      anchorPosition="upRight"
+      panelStyle={{ minWidth: 350 }}
     >
-      <EuiContextMenuPanel className="messageVariablesPanel" items={getMessageVariables()} />
+      <EuiSelectable
+        searchable
+        height={300}
+        data-test-subj={'messageVariablesSelectableList'}
+        isLoading={false}
+        options={shownOptions}
+        listProps={{
+          rowHeight: 70,
+          showIcons: false,
+          paddingSize: 'none',
+          textWrap: 'wrap',
+        }}
+        loadingMessage={i18n.LOADING_VARIABLES}
+        noMatchesMessage={i18n.NO_VARIABLES_FOUND}
+        emptyMessage={i18n.NO_VARIABLES_AVAILABLE}
+        renderOption={renderOption}
+        onChange={(variables) => {
+          variables.map((variable) => {
+            if (variable.checked === 'on' && messageVariablesObject) {
+              onSelectEventHandler(messageVariablesObject[variable.label]);
+            }
+          });
+          setIsVariablesPopoverOpen(false);
+        }}
```

Comment:
@JiaweiWu why do we need to move `onChange` to separate handler? I usually do it if it's pretty big to easily read a code. In that case we can move, but next one is really small. What optimisation do you mean?

Reply:
How I can get rid of `listProps`? I just find this way in docs. What will be an alternative and why it's bad to use it like that?

Reply:
We should try to create separate functions for handlers so they can be wrapped in `useCallback` if we need to later on, inlining the function means every time the component renders, it creates a new reference for that function and the child component will re-render.

> How I can get rid of listProps? I just find this way in docs. What will be an alternative and why it's bad to use it like that?

Sorry I should clarify:

```ts

// Outside of the component 
const listProps = {
  rowHeight: 70,
  showIcons: false,
  paddingSize: 'none',
  textWrap: 'wrap',
}

or 

// Inside of the component if it needs reference to some variable
const listProps = useMemo(() => ({
  rowHeight: 70,
  showIcons: false,
  paddingSize: 'none',
  textWrap: 'wrap',
  someVariable,
}), [someVariable]);

...
<EuiSelectable
  ...
  listProps={listProps}
/>
```

But I think we should just do it for the handlers since the component already has inlined object props before this PR. this is more of a general guideline :slightly_smiling_face: